### DISCUSSION TITLE
[Source generators] Is it possible to add code to different project (not one that is being analyzed)

### DISCUSSION THREAD

**Discussion by maxkoshevoi** [dotnet/roslyn#55755] [Q&A] [answered]
I want to create a Source Generator that would create a unit tests to verify that for all `AutoMapper.Map` calls there's a mapping defined. For this I'll need to analyze  one project for `Map()` calls, and generate unit tests for mapping existence in different project (that contains all other unit tests).

Is this possible to achieve?

**Comment by alrz** [comment, marked as answer]
>I want to create a Source Generator that would create a unit tests to verify that for all AutoMapper.Map calls there's a mapping defined

You could instead create an analyzer to warn/error about those calls that doesn't have a mapping defined.

**Reply by maxkoshevoi** [reply]
This seems like a better approach. But what if mapping is defined is a different project? 
With source generator I could pass all assemblies with mappings as an argument. Can analyzer access all types from referenced projects by it's own?

**Reply by Unskilledcrab** [reply]
Also curious about the answer to this question

**Reply by PathogenDavid** [reply]
> Can analyzer access all types from referenced projects by it's own?

You can access the metadata of referenced projects via `GeneratorExecutionContext.Compilation.References`, but you can not access the source. (Projects are built in isolation, from the compiler's perspective there isn't a difference between a `ProjectReference` and any other reference.)
